python-package/lightgbm/dask.py

```diff
@@ -243,6 +404,24 @@ def _train(
         sum(group) = n_samples.
         For example, if you have a 100-document dataset with ``group = [10, 20, 40, 10, 10, 10]``, that means that you have 6 groups,
         where the first 10 records are in the first group, records 11-30 are in the second group, records 31-70 are in the third group, etc.
+    eval_set : List of (X, y) tuples of Dask data collections, or None, optional (default=None)
+        List of (X, y) tuple pairs to use as validation sets.
+    eval_names: List of strings or None, optional (default=None))
+        Names of eval_set.
+    eval_sample_weight: List of Dask data collections or None, optional (default=None)
+        List of Dask Array or Dask Series, weights for each validation set in eval_set.
+    eval_class_weight: List of dict or str, or None, optional (default=None)
+        List of class weights, one dict or str for each validation set in eval_set.
+    eval_init_score: List of Dask data collections or None, optional (default=None)
+        List of Dask Array or Dask Series, init model score for each validation set in eval_set.
+    eval_group: List of Dask data collections or None, optional (default=None)
+        List of Dask Array or Dask Series, group/query for each validation set in eval_set.
+    eval_metric: String, callable, list or None, optional (default=None)
+        If string, built-in evaluation metric; if callable, a custom evaluation metric.
+        Can also be a list of built-in metric strs and/or custom evaluation metrics.
```

Comment:
Sorry can you please elaborate on why you don't like supporting `eval_metric`? You just mean putting like 99% of the eval-set functionality in this single PR or specifically there's some issue with eval_metric?

Reply:
> specifically there's some issue with eval_metric?

I hope that not 😃 

I meant adding support of `eval_metric` **in this PR.**
We have a separate issue for this functionality specifically: #3956 (and for custom objective too, BTW: #3934).

By addressing multiple issues in a single PR when it is possible not to do so, we have a risk to not give enough attention to some points and miss different things. For example, please refer to my comment above about forgotten note in a docstring.

Reply:
In this case, I specifically asked that supporting a custom eval_metric function be brought in, because it enables writing more rigorous tests that will give us higher confidence in this PR.

https://github.com/microsoft/LightGBM/pull/4101#discussion_r603754561

Reply:
Yeah I totally agree this PR is not in the spirit of "modular commits." This specifically was added after a request - I'm sorry but I need to leave this decision up to you guys on whether to keep! For what it's worth, I don't think supporting custom functional `eval_metric` arguments is the same as supporting custom objective functions (#3934).

Reply:
> In this case, I specifically asked that supporting a custom eval_metric function be brought in

OK, I see.

Let's add custom metric functionality in this PR. But it should be a full support then. I'll keep this in mind when continue my review to check that nothing else related to custom metric is missed.

Reply:
Hey so I replaced docs for `eval_set`, `eval_metric`, and `eval_names` from the sklearn docs. Those for `eval_set` I added a small Note sentence explaining how `evals_result_` and `best_score_` may be impacted by the stochastic choice of returned LightGBM model. Those for `eval_metric` did not need to be altered significantly, as this argument works identically to the non-distributed implementation.